### PR TITLE
[spirv] Move descriptor scalar replacement to legalization stage.

### LINKED ISSUE #3100: [SPIR-V] Global struct variables containing only resources fail validation with '-O0' optimization level
Hi, it seems that the fix to https://github.com/microsoft/DirectXShaderCompiler/issues/2695 doesn't work when optimization is disabled.

### Functional impact
The shader doesn't compile. Cannot use the unoptimized version of the shader for debugging purposes when shader uses the legacy combined texture sampler emulation workaround mentioned in https://github.com/microsoft/DirectXShaderCompiler/issues/2695

### Minimal repro steps
The repro source is the same:

```
struct sampler2D {
  Texture2D t;
  SamplerState s;
};
float4 tex2D(sampler2D x, float2 v) { return x.t.Sample(x.s, v); }

struct v2f {
  float4 vertex : SV_POSITION;
  float2 uv : TEXCOORD0;
};

sampler2D g_CombinedTextureSampler;

float4 PSMain(v2f i) : SV_Target {
  return tex2D(g_CombinedTextureSampler, i.uv);
}
```
Compiling with opt. level 0: `dxc.exe -T ps_6_0 -E PSMain repro_shader.txt -spirv -O0` results in error:

```
fatal error: generated SPIR-V is invalid: UniformConstant OpVariable <id> '8[%g_CombinedTextureSampler]' has illegal type.
From Vulkan spec, section 14.5.2: 
Variables identified with the UniformConstant storage class are used only as handles to refer to opaque resources. Such variables must be typed as OpTypeImage, OpTypeSampler, OpTypeSampledImage, OpTypeAccelerationStructureNV, OpTypeAccelerationStructureKHR, OpTypeRayQueryProvisionalKHR, or an array of one of these types.
  %g_CombinedTextureSampler = OpVariable %_ptr_UniformConstant_sampler2D UniformConstant

note: please file a bug report on https://github.com/Microsoft/DirectXShaderCompiler/issues with source code if possible
```

### Further technical details
Used latest DXC on Windows built from 530958e4b7a2c198f012c8f1dc1656c9dbc02148

### PR DESCRIPTION
**Pull request by ehsannas**
Fixes #3100

### CODE REVIEW

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 1]
```diff
@@ -11937,15 +11937,6 @@ bool SpirvEmitter::spirvToolsOptimize(std::vector<uint32_t> *mod,
     // Add performance passes.
     optimizer.RegisterPerformancePasses();
 
-    // Add flattening of resources if needed.
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp:11940]: Is it ok that we do not apply this when `-O3` option is given?

Comment (ehsannas) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp:11940]: it _will_ be run with `-O3` because `-O3` requires legalization.

Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp:11940]: Do you mean `-O3` will set [`needsLegalization`](https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L591) true?

Comment (ehsannas) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp:11940]: no. Sorry, maybe I didn't explain it properly.

Note that `needsLegalization` is always true if the SPIR-V backend finds a global struct of resources. See [here](https://github.com/microsoft/DirectXShaderCompiler/blob/f1f60648df03f20b1281bfbfbd356c4f6d9f3b48/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L594).

What I mean is: There is no way to run optimizations without legalization. There are 3 code-paths:
1) don't legalize and don't optimize
2) legalize, but don't optimize
3) legalize and optimize

In the past, we were running "descriptor scalar replacement" during optimization. Now we want to do it during legalization.

For  path 1: didn't run the pass before my change, and won't run it now
For  path 2: didn't run the pass before my change, but will run it now --> which will result in the behavior that the user wants in issue #3100 
For  path 3: Ran the pass before my change, and will run it with my change too.

Does it make sense?

Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp:11940]: Gotcha.

**Review by jaebaek — APPROVED** [review]

**Pull request merged**